### PR TITLE
ENH: improve stats.kde density estimation near boundaries

### PR DESCRIPTION
**Pull request by aloctavodia**
When estimating a distribution that has boundaries like a uniform distribution or an exponential, or a beta(0.5, 0.5), etc a KDE has problems estimating the density close to that boundaries. There are many methods to deal with this problem. I have implemented the reflection method.

### CODE REVIEW

**Comment by pv** [comment]
Thanks, but note that adding tests would also be needed here.

**Comment by aloctavodia** [comment]
There is already a test that compares the KDE to a Gaussian distribution, should I add a test that compares to a bounded distribution, like the uniform?

**Comment by pv** [comment]
The test should test the new feature added, iow, be something that 
passes only if the changes here are correct.

**Comment by josef-pkt** [comment]
AFAICS, this uses the data endpoints as boundary points

This cannot be added in this form. It has to be conditional on a user defined option, otherwise it will break the current behavior where we want to extend the normal distribution (gaussian kernel) extended to the entire space.

**Comment by aloctavodia** [comment]
ok, thanks for the input.

**Comment by josef-pkt** [comment]
(sorry if I sound unfriendly, I'm busy and just wanted to get a quick comment in)

**Comment by josef-pkt** [comment]
Does this work for multivariate data? i.e. can the bounds be vectors?

If not, then I think it should raise an exception if the bounds are specified and the data ndim is larger than 1, and specify behavior in docstring.
If yes, then it needs unit test for a multivariate case.

Also, new options need to be added to the docstring.

BTW: statsmodels has kde with bounded support waiting in a PR.
I never looked the the details for boundaries, but it is useful in many cases.

**Comment by aloctavodia** [comment]
I did not check for multivariate data, I think should work, probably we need to check if the boundaries are row or columns vector. I will check it ASAP.

File: scipy/stats/kde.py [hunk 1]
```diff
@@ -170,14 +171,19 @@ def __init__(self, dataset, bw_method=None):
         self.d, self.n = self.dataset.shape
         self.set_bandwidth(bw_method=bw_method)
 
-    def evaluate(self, points):
-        """Evaluate the estimated pdf on a set of points.
+    def evaluate(self, points, lb=None, ub=None):
+        """Evaluate the estimated pdf on a set of points using a Gaussian
+        Kernel, boundary effects corrected using the reflection method.
```
Comment (ev-br) [inline scipy/stats/kde.py]: This will break in the layout of the web docs, where the first line is pulled into the listings, see e.g. http://docs.scipy.org/doc/scipy-0.17.0/reference/generated/scipy.stats.gaussian_kde.html#scipy.stats.gaussian_kde.
 The docstring format should be: the summary line, blank line, the rest of the docstring.  Ideally, keep the lines short enough so that the doc builder does not clip them.

File: scipy/stats/kde.py [hunk 2]
```diff
@@ -170,14 +171,19 @@ def __init__(self, dataset, bw_method=None):
...
 
         Parameters
         ----------
         points : (# of dimensions, # of points)-array
             Alternatively, a (# of dimensions,) vector can be passed in and
             treated as a single point.
+        lb : Lower boundary for the estimated pdf, integer for unidimensional
+        distributions, array_like for multivariate ones
```
Comment (ev-br) [inline scipy/stats/kde.py]:
```
lb: array_like
     Lower boundary ....
```

or `float or array_like` if floats are accepted.

File: scipy/stats/kde.py [hunk 3]
```diff
@@ -191,6 +197,16 @@ def evaluate(self, points):
 
         """
         points = atleast_2d(points)
+        if lb is not None:
+            lb = atleast_2d(lb)
+            f, s = lb.shape
+            if s > f:
+                lb = lb.T
```
Comment (ev-br) [inline scipy/stats/kde.py]: why is this needed?

File: scipy/stats/kde.py [hunk 4]
```diff
@@ -204,23 +220,45 @@ def evaluate(self, points):
...
-                result = result + exp(-energy)
+                diff_0 = self.dataset[:, i, newaxis] - points
+                tdiff_0 = dot(self.inv_cov, diff_0)
+                energy_0 = sum(diff_0 * tdiff_0, axis=0) / 2.0
+                result = result + exp(-energy_0)
+                if lb is not None:
+                    diff_l = self.dataset + points[:, i, newaxis] - 2*lb
+                    tdiff_l = dot(self.inv_cov, diff_l)
+                    energy_l = sum(diff_l * tdiff_l, axis=0) / 2.0
+                    result_l[i] = sum(exp(-energy_l))
```
Comment (ev-br) [inline scipy/stats/kde.py]: Do you happen to have a reference (ideally, add one to the docstring's References section).

**Comment by ev-br** [comment]
I think this should also have a test for something with boundaries. Maybe a data set drawn from a truncated normal distribution (`stats.truncnorm`)

File: scipy/stats/kde.py (hunk 2 above)
Comment (rkern) [inline scipy/stats/kde.py]: Nit: the space before the `:` is required.

File: scipy/stats/kde.py (hunk 4 above)
Comment (aloctavodia) [inline scipy/stats/kde.py]: The reference is already there, ref number 2 "Density Estimation for Statistics and Data Analysis"

File: scipy/stats/kde.py (hunk 3 above)
Comment (aloctavodia) [inline scipy/stats/kde.py]: I add this for multivariate data

File: scipy/stats/kde.py (hunk 2 above)
Comment (aloctavodia) [inline scipy/stats/kde.py]: Yeap, "integer" is typo sorry about that and for the sloppy docstring format. I will fix it ASAP. Thanks for the comments.

**Comment by aloctavodia** [comment]
sorry about that a previous commit had test for boundaries before, but then I accidentally removed it.

**Comment by aloctavodia** [comment]
Is this OK to be merged or further improvements are needed?

**Comment by ev-br** [comment]
Sorry this fell offthe radar. @josef-pkt are you OK with this in its current form?

**Comment by josef-pkt** [comment]
overall looks fine to me, (I didn't figure out the math (which is not obvious to me just from reading the code)

one question: What happens if the user specifies bounds that are inside the sample, i.e. there are out of bounds points?

My first impression was that the bounds, lb and ub, should be an argument in `__init__` instead of the individual methods, but because there is no related state, precalculated attributes, the current version looks fine to me. If bounds checking on the sample is needed, then it would be better to do it only once in `__init__`.

Is the `_norm_factor` still correct for the normalization? 
It's possible that it is because reflection just shifts the weights/probability, and that's likely based on the first unit test.

small correction: "interval" has only one `l` in english not two (in unit tests)

File: scipy/stats/kde.py [hunk 5]
```diff
@@ -204,23 +219,45 @@ def evaluate(self, points):
...
-        result = result / self._norm_factor
+                diff_0 = self.dataset - points[:, i, newaxis]
+                tdiff_0 = dot(self.inv_cov, diff_0)
+                energy_0 = sum(diff_0 * tdiff_0, axis=0) / 2.0
+                result[i] = sum(exp(-energy_0), axis=0)
+                if lb is not None:
+                    diff_l = self.dataset + points[:, i, newaxis] - 2*lb
+                    tdiff_l = dot(self.inv_cov, diff_l)
+                    energy_l = sum(diff_l * tdiff_l, axis=0) / 2.0
+                    result_l[i] = sum(exp(-energy_l), axis=0)
```
Comment (josef-pkt) [inline scipy/stats/kde.py]: one possible change: this could be added inplace to `result[i]` instead of using additional temporary arrays, AFAICS.

**Comment by aloctavodia** [comment]
Thanks for the input. I will check this ASAP.

**Comment by aloctavodia** [comment]
The error CircleCi reports seems to be unrelated to the PR.

**Comment by ev-br** [comment]
@aloctavodia Would you be able to rebase this so that it can go into 0.19.0?

**Comment by ev-br** [comment]
Needs a rebase and a review.
Won't make it to 0.19.0, moving the milestone to 1.0

**Comment by aloctavodia** [comment]
Sorry for the late reply, at the moment I am very busy with other things. There have been a long time since my first commit. I think this is a useful feature but now I am unsure about this particular implementation. One of the reasons is that it will be slower than the current version. I have implemented a "fast KDE" version into PyMC3 but is not general enough to be used in SciPy. In summary, this implementation is not good enough for SciPy.

**Comment by rgommers** [comment]
Thanks for the update and the attempt at improving this functionality @aloctavodia. If you think this PR is not good enough, let's close it for now.
